pydantic_graph/pydantic_graph/state.py

```diff
@@ -0,0 +1,133 @@
+from __future__ import annotations as _annotations
+
+import copy
+from collections.abc import Sequence
+from contextvars import ContextVar
+from dataclasses import dataclass, field
+from datetime import datetime
+from typing import Annotated, Any, Callable, Generic, Literal, Union
+
+import pydantic
+from pydantic_core import core_schema
+from typing_extensions import TypeVar
+
+from . import _utils
+from .nodes import BaseNode, End, RunEndT
+
+__all__ = 'StateT', 'NodeStep', 'EndStep', 'HistoryStep', 'deep_copy_state'
+
+
+StateT = TypeVar('StateT', default=None)
+"""Type variable for the state in a graph."""
+
+
+def deep_copy_state(state: StateT) -> StateT:
+    """Default method for snapshotting the state in a graph run, uses [`copy.deepcopy`][copy.deepcopy]."""
+    if state is None:
+        return state
+    else:
+        return copy.deepcopy(state)
+
+
+@dataclass
+class NodeStep(Generic[StateT, RunEndT]):
+    """History step describing the execution of a node in a graph."""
+
+    state: StateT
+    """The state of the graph before the node is run."""
+    # node: Annotated[BaseNode[StateT, RunEndT], pydantic.WrapSerializer(node_serializer), pydantic.PlainValidator(node_validator)]
+    node: Annotated[BaseNode[StateT, RunEndT], CustomNodeSchema()]
+    """The node that was run."""
+    start_ts: datetime = field(default_factory=_utils.now_utc)
+    """The timestamp when the node started running."""
+    duration: float | None = None
+    """The duration of the node run in seconds."""
+    kind: Literal['node'] = 'node'
+    """The kind of history step, can be used as a discriminator when deserializing history."""
+    # waiting for https://github.com/pydantic/pydantic/issues/11264, should in InitVar
+    snapshot_state: Annotated[Callable[[StateT], StateT], pydantic.Field(exclude=True, repr=False)] = field(
+        default=deep_copy_state, repr=False
+    )
+    """Function to snapshot the state of the graph."""
```

Comment:
I feel this deserves a bit more documentation about why you might want to override the value. In particular, adding a note like "uses copy.deepcopy by default; you may want to provide a different implementation if your state object is memory intensive and largely immutable" (or similar)

Reply:
ye agreed, I think this can come later.